### PR TITLE
wip: paid subscriptions system and articles

### PR DESCRIPTION
**Pull request by DioFun**
Le but est d'avoir un suivi et des correctifs aux fur et à mesure pour le moment quant au système d'abonnement payant.
Développement de la gestion des articles, remboursements et abonnements payants.

### CODE REVIEW

**Comment by D0gmaDev** [comment]
Links to #455

File: app/controllers/admin/dashboard_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Admin
+  class DashboardController < ApplicationController
+    def index
+      authorize! :manage, :all
+      @articles = Article.where(deleted_at: nil).order(created_at: :desc)
+      @subscription_offers = SubscriptionOffer.where(deleted_at: nil).order(created_at: :desc)
+      @payment_methods = PaymentMethod.where(deleted_at: nil).order(created_at: :desc)
```
Comment (Letiste) [inline app/controllers/admin/dashboard_controller.rb:9]: nit: we could move the logic of filtering deleted records to the models and use [scopes](https://guides.rubyonrails.org/active_record_querying.html#scopes), maybe use a default scope?

File: app/models/user.rb [hunk 2]
```diff
@@ -2,8 +2,11 @@
 
 class User < ApplicationRecord
   has_many :machines, dependent: :destroy
-  has_many :subscriptions, dependent: :destroy
   has_many :free_accesses, dependent: :destroy
+  has_many :sales_as_client, class_name: 'Sale', foreign_key: 'client_id', dependent: :destroy, inverse_of: :client
+  has_many :sales_as_seller, class_name: 'Sale', foreign_key: 'seller_id', dependent: :nullify, inverse_of: :seller
+  has_many :refunds, foreign_key: 'refunder_id', dependent: :destroy, inverse_of: :refunder
+  has_many :subscriptions, through: :sales_as_client, dependent: :destroy
```
Comment (Letiste) [inline app/models/user.rb:9]: Do we need to keep anything regarding legal/financial when a user is deleted?

File: test/controllers/admin/articles_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      end
+
+      assert_template 'admin/articles/new'
+    end
+
+    test 'should not destroy article if soft_delete' do
+      assert_no_difference 'Article.count' do
+        @articles.soft_delete
+      end
+    end
```
Comment (Letiste) [inline test/controllers/admin/articles_controller_test.rb:37]: nit: we are testing the logic of the model here, it would be better to put it in the model tests

File: test/models/article_test.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  test 'price should be positive' do
+    @subscription_offer.price = -5
+    assert_not_predicate @subscription_offer, :valid?
+  end
+
+  test 'article should soft delete' do
+    assert_no_difference 'Article.count' do
+      @subscription_offer.soft_delete
+    end
```
Comment (Letiste) [inline test/models/article_test.rb:37]: would it be a more thorough test if we checked that the deleted_at field is nil before the soft delete and not nil after? As we have some logic requiring the deleted_at to be not nil

File: test/models/refund_test.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    assert_difference 'Refund.count', -1 do
+      @refund.destroy
+    end
+  end
+
+  test 'destroy refun should destroy refunds_subscription_offers' do
+    assert_difference 'Refund.count', -1 do
+      @refund.destroy
+    end
+  end
```
Comment (Letiste) [inline test/models/refund_test.rb:20]: I don't think the tests are doing what their names are implying. We are testing that `@refund.destroy` deletes the refund, not the associated records

File: app/models/sale.rb [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  def update_total_price
+    total = 0
+    articles_sales.each do |rec|
+      total += rec.quantity * Article.find(rec.article_id).price
+    end
+    sales_subscription_offers.each do |rec|
+      total += rec.quantity * SubscriptionOffer.find(rec.subscription_offer.id).price
+    end
+    self.total_price = total
+  end
```
Comment (Letiste) [inline app/models/sale.rb:43]: it's a bit of a code smell here because we need to call `update_total_price` before getting the total price. we could compute this value every time we get the total price instead, we would have the guarantee always to have the correct total price.
one benefit of being explicit on the update of the total price is that we don't compute it every time we get the total price but we don't need this kind of performance optimization (and we would need to adapt the logic anyway to guarantee that the total price is always up-to-date)

**Review by Letiste — COMMENTED** [review]
First batch of comments, I didn't finish the review yet. I didn't review the views because the focus here is more on the sales logic than adding some forms. I still need to review the sale/refund logic + the invoice generation

**Review by nymous — COMMENTED** [review]
Same, just posting what I wrote for now (also you pushed after I started to review so some comments might be outdated ^^')

File: app/controllers/admin/dashboard_controller.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Admin
+  class DashboardController < ApplicationController
+    def index
+      authorize! :manage, :all
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/controllers/admin/dashboard_controller.rb:6]: That is temporary permission for now

File: app/models/article.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+class Article < ApplicationRecord
+  has_many :articles_sales, dependent: :restrict_with_error
+  has_many :sales, through: :articles_sales
+  has_many :articles_refunds, dependent: :restrict_with_error
+  has_many :refunds, through: :articles_refunds
+
+  validates :name, presence: true, allow_blank: false
+  validates :price, presence: true, allow_blank: false,
+                    numericality: { greater_than_or_equal_to: 0, only_integer: true, message: 'Must be a positive
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/article.rb:11]: It's something that should be discuss. In a near or far future if we want to manage stocks of products maybe we'd like to sell things for free.

File: app/models/article.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+                     number. Maximum 2 numbers after comma' }
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/article.rb:12]: In the UI, the user will be ask to enter a decimal number and then it'll be multiply by 100

[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/article.rb:12]: In the UI, the user will be ask to enter a decimal number and then it'll be multiply by 100

File: app/models/invoice.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class Invoice < ApplicationRecord
+  has_one :sale, dependent: :restrict_with_exception
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/invoice.rb:4]: The place we use restrict_with_error is because when we try to delete the record we don't know if it will fail or not and if it fail we soft delete it. But here, it's not planned to be deleted in the code so it's with an exception.

File: app/models/sale.rb (hunk 6 above)
Comment (DioFun) [reply app/models/sale.rb:43]: We call for `update_total_price` when there is no subscription offers nor articles to be add to the sale so there will not have any update of the price after. I agree that is not quite convincing but I did not find anything else to do so.

File: app/models/user.rb (hunk 2 above)
Comment (DioFun) [reply app/models/user.rb:9]: We only need to keep the invoices for ten years which are not deleted when destroying a user.

File: db/migrate/20240720124420_create_sale_article_details.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    create_table :articles_sales do |t|
+      t.references :sale, null: false, foreign_key: true
+      t.references :article, null: false, foreign_key: true
+      t.integer :quantity
+
+      t.timestamps
+    end
+    # create_join_table :sales, :articles, column_options: { foreign_key: true } do |t|
+    #   t.integer :quantity, null: false
+    # end
```
[the remark this answers is not in the document]
Comment (DioFun) [reply db/migrate/20240720124420_create_sale_article_details.rb:14]: At first, we tried to use the `create_join_table` method but that do not create a primary, auto-increment key thus when we tried to delete a record from the join table it was a fail. So I decided to switch to a classic table creation with `create_table` in order to have such a key.

File: app/controllers/sales_controller.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+      flash.now[:error] = "You can't process an empty sale!"
+      return render :new, status: :unprocessable_entity
+    end
+    @sale.verify if @sale.payment_method.auto_verify
+    @sale.gen_temp_invoice
+    authorize! :create, @sale
+    if @sale.save
+      name = @sale.generate_invoice_id
+      @sale.invoice.pdf.attach(io: @sale.invoice.generate, filename: name, content_type: 'application/pdf')
+      @sale.save!
```
[the remark this answers is not in the document]
Comment (D0gmaDev) [reply app/controllers/sales_controller.rb]: I believe this is related to the fact that we need to know the id of the invoice to store it in the invoice json itself

File: db/migrate/20240720124610_create_active_storage_tables.active_storage.rb
[the remark this answers is not in the document]
Comment (DioFun) [reply db/migrate/20240720124610_create_active_storage_tables.active_storage.rb:1]: Yes we want to store less than 100 MB of pdf (pdf are very light)

File: app/controllers/sales_controller.rb [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    @articles = Article.where(deleted_at: nil)
+    @subscription_offers = SubscriptionOffer.where(deleted_at: nil).order(duration: :desc)
+    @payment_methods = PaymentMethod.where(deleted_at: nil)
+    authorize! :new, @sale
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def create
+    @sale = @owner.sales_as_client.new(reformated_params)
+    @sale.update_total_price
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/controllers/sales_controller.rb]: Has been moved but not sure if it's enough

File: app/models/article.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (Letiste) [reply app/models/article.rb:11]: If we are not sure it's something we need now, I would suggest allowing a price only strictly greater than 0. I think it would simplify some logic (not handling division by zero and a total price of zero). And that won't block us from adapting the logic in the future to support a price of 0

File: app/models/sale.rb (hunk 6 above)
Comment (Letiste) [reply app/models/sale.rb:43]: Do we need to have the total price in the database? The total price can be calculated at runtime by summing the price of articles and subscriptions, keeping it as a column in the database is redundant and we might have consistency issues (the total price of the sale is different from the sum of the articles' and subscriptions' price because we forgot to update the total price somewhere). And we can change `update_total_price` to be a getter (renaming it to `total_price` and returning the computed total price)

File: db/migrate/20240720124610_create_active_storage_tables.active_storage.rb
[the remark this answers is not in the document]
Comment (Letiste) [reply db/migrate/20240720124610_create_active_storage_tables.active_storage.rb:1]: What happens if we lose the disk? Do we have any backup mechanism in place to restore the pdfs (or regenerate them)?

File: app/models/article.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (Letiste) [reply app/models/article.rb:12]: Ah the price is in cents, I didn't catch that, makes sense! So we have two different validations:
- in the backend, we only expect integers
- in the frontend, we only expect float with a maximum of 2 decimals

In this case, I would change the error message in the backend to something like `Must be a positive integer` because it shouldn't be aware of how this integer is computed in the frontend. And have some logic in the frontend to display an error when the price is more than 2 decimals. What do you think?

File: app/models/invoice.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (Letiste) [reply app/models/invoice.rb:4]: Got it! I think it's worth adding some comments on that, and some tests on the `destroy` code path (not only `soft_delete`). Would use `restrict_with_exception` everywhere and explicitly handle the exception and soft delete when it happens be more explicit? Not sure if it's very ruby-like though, I will defer to @nymous for this

[the remark this answers is not in the document]
Comment (Letiste) [reply app/models/invoice.rb:4]: Thinking about it again, if it's expected to try to delete an article but we should block the deletion when it's currently used by a sale, having an error instead of an exception makes more sense

File: app/models/sale.rb (hunk 6 above)
Comment (DioFun) [reply app/models/sale.rb:43]: The fact is each time we would like to show a sale we need to calculate again the total price of the sale. That would be redundant because once a sale is created we want it to be immutable so after the creation the price couldn't be changed so there is no need to update the price later we just have to calculate it when we create it and it enables to avoid recalculate it each time we want to display which is the case when we show a user, we'd like to show every sales and their total price.

File: db/migrate/20240720124610_create_active_storage_tables.active_storage.rb
[the remark this answers is not in the document]
Comment (D0gmaDev) [reply db/migrate/20240720124610_create_active_storage_tables.active_storage.rb:1]: bad things I guess. The pdf can be regenerated using the json stored in the invoice object, but in case we also lose those, i think it would be a good idea to also send the invoices in a folder on another server when we create them

File: app/models/invoice.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/invoice.rb:4]: Yes that has been raised by @nymous in one of his comment

File: app/models/article.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/article.rb:12]: Yes I agree but I'm not sure on how to deal with that to change the error in front.

File: db/migrate/20240720124610_create_active_storage_tables.active_storage.rb
[the remark this answers is not in the document]
Comment (DioFun) [reply db/migrate/20240720124610_create_active_storage_tables.active_storage.rb:1]: It would surely be host on the PVE which is backed up but that are consideration when we will deploy it.

File: app/models/sale.rb (hunk 6 above)
Comment (Letiste) [reply app/models/sale.rb:43]: I don't think recalculating it every time is an issue. We certainly don't need that kind of optimization (we wouldn't even notice it). Making the code more resilient to bugs is more important in my opinion.

Regarding storage issues, we have a redundancy with the pdf if we need to know the price. I would say it's good enough to make sure we can't lose the invoices.

My comment is not blocking in any way, I don't mind keeping it like this

File: app/models/sale.rb [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  def compute_total_price
+    total = 0
+    articles_sales.each do |rec|
+      total += rec.quantity * Article.find(rec.article_id).price
+    end
+    sales_subscription_offers.each do |rec|
+      total += rec.quantity * SubscriptionOffer.find(rec.subscription_offer.id).price
+    end
+    total
+  end
```
Comment (Letiste) [inline app/models/sale.rb:45]: We don't care about that at our scale because we are talking about one or two articles, but I think here we would make a request to the database per article/subscription. That can be a problem once you get to dozens of articles.
We don't need to change anything for our case though, it's simpler like that

File: app/models/setting.rb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class Setting < ApplicationRecord
```
[the remark this answers is not in the document]
Comment (D0gmaDev) [reply app/models/setting.rb:3]: @Letiste I step in here, we need full reliable control over the id of the next invoice we want to generate, for the invoices to be legaly compliant, in case a part of the db is cleared, or we need to purposly skip ids to account for manual invoices, or just the fact that we have to start the id at 7000-ish to ensure continuation from re2o. 
Therefore the generated id is not flexible enough.

File: app/models/articles_sale.rb [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+  private
+
+  def consolidate_duplication
+    duplicate = ArticlesSale.where(article_id: article_id, sale_id: sale_id).where.not(id: id).first
+    return unless duplicate
+
+    errors.add(:base, 'Please merge the quantities of the same articles !')
+    throw :abort
+  end
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/articles_sale.rb:19]: Here, we want to avoid different articles_sales record for the same sale and article. Example: the user add 2 articles but these 2 are the same article.

File: app/models/sale.rb [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+      if quantity.positive?
+        sales_subscription_offers.new(subscription_offer_id: offer.id, quantity: quantity)
+        duration -= quantity * offer.duration
+      end
+    end
+    return true if duration.zero?
+
+    errors.add(:base, 'Subscription offers are not exhaustive!')
+    false
+  end
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/sale.rb:79]: This function aims to generate the subiscription_offer for a given duration, duration is the number of months the subscription last. Example : an adherent want to pay for 13 months and we have two different offer 12 months is 50€ and 1 month is 5€ then it will create a sales_subscription_offers of 12 months and of 1 months. So we leave the loop when reach 0 because we do not need to continue as we already split the sale into the different offers we have. Lastly, we return false, because it's quite obscure the error process and the throw in my case i'd like just to send a message to the user not crash the website.

File: app/models/subscription.rb [hunk 18]
```diff
@@ -1,22 +1,26 @@
...
-    return if cancelled_at_was.nil?
+  # private
 
-    errors.add(:cancelled_at, 'Subscription has already been cancelled')
-  end
+  # def cannot_change_after_cancelled
+  #   return if cancelled_at_was.nil?
+  #
+  #   errors.add(:cancelled_at, 'Subscription has already been cancelled')
+  # end
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/subscription.rb:25]: I just commented it out for now because we would surely use it again when we'll make the refund module.

File: test/services/invoice_pdf_generator_test.rb
Comment (nymous) [inline test/services/invoice_pdf_generator_test.rb:1]: Wow very nice tests on a hard-to-test feature!

**Review by nymous — CHANGES_REQUESTED** [review]

**Comment by nymous** [comment]
Just to show you what the errors look like in the reworked sale form:
image
The duplicate articles are shown in red

File: app/models/article.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (nymous) [reply app/models/article.rb:12]: See #478

[the remark this answers is not in the document]
Comment (nymous) [reply app/models/article.rb:12]: See #478

File: app/models/article.rb [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class Article < ApplicationRecord
+  has_many :articles_sales, dependent: :restrict_with_error
+  has_many :sales, through: :articles_sales
+  has_many :articles_refunds, dependent: :restrict_with_error
+  has_many :refunds, through: :articles_refunds
+
+  validates :name, presence: true, allow_blank: false
+  validates :price, presence: true, allow_blank: false,
```
[the remark this answers is not in the document]
Comment (nymous) [reply app/models/article.rb:10]: See #478

File: app/controllers/sales_controller.rb [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    @sale.articles_sales.new
+    @articles = Article.all
+    @subscription_offers = SubscriptionOffer.order(duration: :desc)
+    @payment_methods = PaymentMethod.all
+    authorize! :new, @sale
+  end
+
+  def create
+    @sale = @owner.sales_as_client.new(sales_params)
+    unless @sale.generate(duration: params[:sale][:duration], seller: current_user)
```
[the remark this answers is not in the document]
Comment (nymous) [reply app/controllers/sales_controller.rb]: See #479

File: app/views/admin/dashboard/index.html.erb [hunk 21]
```diff
@@ -0,0 +1,36 @@
+<%# locals: () %>
+
+<h1>Admin dashboard</h1>
+
+<% if can?(:read, Article) %>
+  <h2>Articles</h2>
+  <% if can?(:create, Article) %>
+    <%= button_to new_article_path, method: :get, class: 'button-primary' do %>
```
[the remark this answers is not in the document]
Comment (nymous) [reply app/views/admin/dashboard/index.html.erb:8]: See #482

File: app/models/payment_method.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class PaymentMethod < ApplicationRecord
+  has_many :sales, dependent: :restrict_with_error
+  has_many :refunds, foreign_key: 'refund_method_id', dependent: :restrict_with_error, inverse_of: :refund_method
+
+  validates :name, presence: true, allow_blank: false
+  validates :auto_verify, inclusion: { in: [true, false] }
+
+  default_scope { where(deleted_at: nil) }
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/payment_method.rb:10]: See #480

File: app/models/subscription_offer.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+  has_many :refunds_subscription_offers, dependent: :restrict_with_error
+  has_many :refunds, through: :refunds_subscription_offers
+
+  validates :duration, presence: true, allow_blank: false,
+                       numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+  validates :price, presence: true, allow_blank: false,
+                    numericality: { greater_than_or_equal_to: 0, only_integer: true, message: 'Must be a positive
+                     number. Maximum 2 numbers after comma' }
+
+  default_scope { where(deleted_at: nil) }
```
[the remark this answers is not in the document]
Comment (DioFun) [reply app/models/subscription_offer.rb:15]: See #480

File: test/models/invoice_test.rb [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+
+  test 'should create invoice with pdf' do
+    @invoice.generation_json = @invoice.send(:generate_hash, @sale).to_json
+
+    assert_difference('ActiveStorage::Attachment.count', 1) do
+      @invoice.send(:create_invoice)
+      @invoice.save!
+    end
+
+    assert_predicate @invoice.pdf, :attached?
```
[the remark this answers is not in the document]
Comment (nymous) [reply test/models/invoice_test.rb:52]: See #483

File: test/controllers/admin/subscription_offers_controller_test.rb [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+      end
+
+      assert_template 'admin/subscription_offers/new'
+    end
+
+    test 'should not destroy offer if soft_delete' do
+      assert_no_difference 'SubscriptionOffer.unscoped.count' do
+        @subscription_offer.soft_delete
+      end
+    end
```
[the remark this answers is not in the document]
Comment (nymous) [reply test/controllers/admin/subscription_offers_controller_test.rb:37]: There was already a model test for this, let's remove the controller test

File: app/models/invoice.rb [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
```
[the remark this answers is not in the document]
Comment (nymous) [reply app/models/invoice.rb:5]: I think I understood why it's this way: the invoice can be linked to either a sale or a refund, so we need the reference to be in the sale/refund table (the alternative would be to use Polymorphic relation with an "Invoiceable", but it's more complex and might be less supported by the framework or other dependencies)

**Review by nymous — APPROVED** [review]
FINALLY
